cpp/src/groupby/new_groupby.cu

```diff
@@ -0,0 +1,192 @@
+   * GDF_COLUMN_SIZE_MISMATCH if the columns are not of equal length, 
+   */
+  /* ----------------------------------------------------------------------------*/
+  gdf_error verify_columns(gdf_column * cols[], int num_cols)
+  {
+    GDF_REQUIRE((nullptr != cols[0]), GDF_DATASET_EMPTY);
+
+    gdf_size_type const required_size{cols[0]->size};
+
+    for(int i = 0; i < num_cols; ++i)
+    {
+      GDF_REQUIRE(nullptr != cols[i], GDF_DATASET_EMPTY); 
+      GDF_REQUIRE(nullptr != cols[i]->data, GDF_DATASET_EMPTY);
+      GDF_REQUIRE(required_size == cols[i]->size, GDF_COLUMN_SIZE_MISMATCH );
+
+      // TODO Remove when null support for hash-based groupby is added
+      GDF_REQUIRE(nullptr == cols[i]->valid || 0 == cols[i]->null_count, GDF_VALIDITY_UNSUPPORTED);
+    }
+    return GDF_SUCCESS;
+  }
+} // anonymous namespace
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief  Groupby operation for an arbitrary number of key columns and an
+ * arbitrary number of aggregation columns.
+ *
+ * "Groupby" is a reduce-by-key operation where rows in one or more "key" columns
+ * act as the keys and one or more "aggregation" columns hold the values that will
+ * be reduced. 
+ *
+ * The output of the operation is the set of key columns that hold all the unique keys
+ * from the input key columns and a set of aggregation columns that hold the specified
+ * reduction among all identical keys.
+ * 
+ * @Param[in] in_key_columns[] The input key columns
+ * @Param[in] num_key_columns The number of input columns to groupby
+ * @Param[in] in_aggregation_columns[] The columns that will be aggregated
+ * @Param[in] num_aggregation_columns The number of columns that will be aggregated
+ * @Param[in] agg_ops[] The aggregation operations to perform. The number of aggregation
+ * operations must be equal to the number of aggregation columns, such that agg_op[i]
+ * will be applied to in_aggregation_columns[i]
+ * @Param[in,out] out_key_columns[] Preallocated buffers for the output key columns
+ * columns
+ * @Param[in,out] out_aggregation_columns[] Preallocated buffers for the output 
+ * aggregation columns
+ * @Param[in] options Structure that controls behavior of groupby operation, i.e.,
+ * sort vs. hash-based implementation, whether or not the output will be sorted,
+ * etc. See definition of gdf_context.
+ * 
+ * @Returns GDF_SUCCESS upon succesful completion. Otherwise appropriate error code
+ */
+/* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_by(gdf_column* in_key_columns[],
+                       int num_key_columns,
+                       gdf_column* in_aggregation_columns[],
+                       int num_aggregation_columns,
+                       gdf_agg_op agg_ops[],
+                       gdf_column* out_key_columns[],
+                       gdf_column* out_aggregation_columns[],
+                       gdf_context* options)
```

Comment:
Still think we should change the name of `gdf_context`... Context implies state.

Reply:
I think renaming it is outside my scope. I generally dislike "options" structs as a whole. But for reference `gdf_context` does technically contain state in that `gdf_context.flag_sorted` indicates if the input data is already sorted or not.